### ISSUE TITLE
Linux, ssh-agent: "sign_and_send_pubkey: signing failed: agent refused operation"

### ISSUE THREAD

**Issue by t-aus-m** [keepassxreboot/keepassxc#2606]
## Expected Behavior
Having activated the ssh-agent feature of keepassxc, signing into an ssh server with any of the provided keys should work without any problem.

## Current Behavior
Before starting keepassxc, there are no identities provided to the ssh-agent.
```
$ ssh-add -l                             
The agent has no identities.
```
After starting keepassxc, there are all identities present, that are expected.
```
$ ssh-add -l                             
2048 SHA256: *something something* *identity name 1* (RSA)
2048 SHA256: *something something* *identity name 2* (RSA)
2048 SHA256: *something something* *identity name 3* (RSA)
```
So if I now try to log into a server that is setup with one of the keys:
```
$ ssh user@example.com
sign_and_send_pubkey: signing failed: agent refused operation
user@server-ip: Permission denied (publickey).
```
I get the same result with two Systems. I'm completely sure that the Keys and their Passwords are correct. When I export them to my local filesystem and manually add them, everything works fine.

## Context
Both systems I had this issue on are running Ubuntu 18.10, I've tried this with 4 different servers from 4 different hosts. The issue persists across all of them.

## Debug Info
KeePassXC - Version 2.3.4
Revision: 6fe821c

Bibliotheken:
- Qt 5.11.1
- libgcrypt 1.8.3

Betriebssystem: Ubuntu 18.10
CPU-Architektur: x86_64
Kernel: linux 4.18.0-13-generic

Aktivierte Erweiterungen:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

**Comment by cl0ne** [comment]
Can confirm on Arch Linux

```
KeePassXC - Версія 2.3.4
Ревізія: 6fe821c

Бібліотеки:
- Qt 5.12.0
- libgcrypt 1.8.4

Операційна система: Arch Linux
Архітектура ЦП: x86_64
Ядро: linux 4.20.0-zen1-1-zen

Увімкнені розширення:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey
```

**Comment by hifi** [comment]
Do you have the require confirmation box checked in KeePassXC?

Technically after the key has been added to the agent successfully KeePassXC is not part of the signing process.

**Comment by t-aus-m** [comment]
I've just rechecked:

> Do you have the require confirmation box checked in KeePassXC?

I have.

> Also, what's the value of your SSH_AUTH_SOCK environment variable?

It does indeed point to /run/user.

**Comment by hifi** [comment]
Try disabling that and if it doesn't help don't use GNOME Keyring as an SSH agent, that's pretty much what I can tell you.

**Comment by t-aus-m** [comment]
Disabling what exactly? The confirmation box? I just tried that and it does not solve the problem.

And how do I not use GNOME Keyring as an SSH-Agent? I have no Idea about that one.

**Comment by Lgmrszd** [comment]
Can confirm this bug. It happens if "Require user confirmation.." checkbox enabled, after disabling it and restarting keepassxc it works fine.

**Comment by hifi** [comment]
If that indeed is the issue then the best we could do is add a warning first time the checkbox is selected that not all agents support it or need extra configuration.

**Comment by t-aus-m** [comment]
For some reason it works now. Looks like some update fixed my issue.

**Comment by roded** [comment]
I've just experienced the same issue, unchecking the "Require user confirmation..." checkbox solved the issue. ssh-agent is running as a user systemctl service.

KeePassXC - Version 2.4.1
Revision: 7bafe65

**Comment by aligator** [comment]
you have to install either x11-ssh-askpass or any other replacement: https://wiki.archlinux.org/index.php/SSH_keys#Alternative_passphrase_dialogs
I use lxqt-openssh-askpass

If you use a replacement, which is not installed to /usr/lib/ssh/x11-ssh-askpass you may have to change the environment variable. 
For this add the following to /etc/environment
`SSH_ASKPASS=/usr/bin/lxqt-openssh-askpass`

(I use Manjaro.)

**Comment by cl0ne** [comment]
[Also](https://github.com/keepassxreboot/keepassxc/issues/1627#issuecomment-463724120):
> Confirmation with ksshaskpass works only when ksshaskpass executable is symlinked to `/usr/lib/ssh/ssh-askpass.`

Arch Linux, KDE

**Comment by idotobi** [comment]
```
$ hostnamectl
  Operating System: Arch Linux
            Kernel: Linux 5.6.8-arch1-1
      Architecture: x86-64
```

🎉 After 3h I finally found the solution to my problem. I'll post it here in case anyone runs in the same issue.

tldr: Adding the environment variable `DISPLAY` to the systemd file in `.config/systemd/user/ssh-agent.service` (which I took from [Arch Wiki](https://wiki.archlinux.org/index.php/SSH_keys#Start_ssh-agent_with_systemd_user)) makes ssh-askpass work as expected.

```
[Unit]
Description=SSH key agent

[Service]
Type=simple
Environment=SSH_AUTH_SOCK=%t/ssh-agent.socket
Environment=DISPLAY=:0
ExecStart=/usr/bin/ssh-agent -D -a $SSH_AUTH_SOCK

[Install]
WantedBy=default.target
```

Long story:

For me the problem was that ssh-askpass did not work right after booting the PC, but only after manually restarting the service via `systemctl --user restart ssh-agent.service`.

By getting the enviornment variables of the ssh-agent process before and after the restart I found that `DISPLAY` was only part of the environment after the restart.

FYI: You can get the environment variables of a running process in a nice format via:

```
ps -aux | grep ssh        # => get $pid 
cat /proc/$pid/environ | tr \\0 \\n
```

If anyone knows if there is a better way than hardcoding `DISPLAY` please let me know.

**Comment by karolyi** [comment]
Added to what @idotobi said, if you use a non-standard askpass binary (not `/usr/lib/ssh/ssh-askpass`), add its path to the unit file as well:
```ini
[Unit]
Description=SSH key agent

[Service]
Type=simple
Environment=DISPLAY=:0
Environment=SSH_AUTH_SOCK=%t/ssh-agent.socket
Environment=SSH_ASKPASS=/usr/lib/seahorse/ssh-askpass
ExecStart=/usr/bin/ssh-agent -D -a $SSH_AUTH_SOCK

[Install]
WantedBy=default.target
```

In my case it's the one from seahorse that is the default password manager for gnome.

Adding the `SSH_ASKPASS` variable is also beneficial to your shell startup script. This is the one for fish shell:

```shell
$ grep SSH_ASKPASS ~/.config/fish/config.fish 
set --export SSH_ASKPASS /usr/lib/seahorse/ssh-askpass

```

**Comment by Alethephobe** [comment]
`$ hostnamectl  
Operating System: Debian GNU/Linux 12 (bookworm)
Kernel: Linux 6.1.0-18-amd64  
Architecture: x86-64  
`
As a linux beginner I don't understand why the issue is closed even tho it still persists.

The issue vanishes, when I give up on the confirmation box, but since I would like that, I tried the following:

1. install ssh-askpass
- sudo apt-get install ssh-askpass

2. adding environment path
- nano ~/.bashrc
- addding "export SSH_ASKPASS=/usr/bin/ssh-askpass"
- apply changes: source ~/.bashrc

3. check "echo $DISPLAY" returns :0

After this + relogging the issue persists. 
Excuse me If i missed something important, I would appreciate any guidance.

**Comment by de-es** [comment]
idotobi's service unit solution was working for quite some time, but unfortunately not anymore for me. It looks like the `DISPLAY` env var alone is not enough. I figured the following working solution.

`~/.config/systemd/user/ssh-agent.service`:

```
# Requires SSH_AUTH_SOCK="$XDG_RUNTIME_DIR/ssh-agent.socket" to be set in environment
[Unit]
Description=OpenSSH key agent
Documentation=man:ssh-agent(1) man:ssh-add(1) man:ssh(1)
ConditionEnvironment=!SSH_AGENT_PID
PartOf=graphical-session.target

[Service]
Environment=SSH_AUTH_SOCK=%t/ssh-agent.socket
Environment=SSH_ASKPASS=/usr/bin/ksshaskpass
ExecStart=/usr/bin/ssh-agent -D -a $SSH_AUTH_SOCK
PassEnvironment=SSH_AGENT_PID
SuccessExitStatus=2
Type=simple

[Install]
WantedBy=graphical-session.target
```

(Manjaro KDE with ksshaskpass)

**Comment by ErtaiBlue** [comment]
> If anyone knows if there is a better way than hardcoding `DISPLAY` please let me know.

to add upon what @idotobi said

I also had this issue of ssh-agent failing to sign because it didn't have DISPLAY set. I was starting ssh-agent in my .zprofile rather than using a systemd unit. Because my zprofile last statement is launching the graphical environment, ssh-agent didn't get the DISPLAY environment variable set. Moving ssh-agent launch from .zprofile to .xinitrc fixed it.
Lines to use in .xinitrc or .bashrc can be found in arch wiki: https://wiki.archlinux.org/title/SSH_keys#SSH_agents
Hope that helps someone else.
